Add powTarget and other changes for Whisper RPC implementation

kdeme opened the pull request:
PR with some changes for upcoming Whisper RPC PR.

Biggest change is adding the option of `powTarget` so that you don't always need to wait for `powTime`. This is an option in the `shh_post` RPC call.

Annoyingly, this does violate the Whisper EIP in its current implementation (so does geth by the way, I understand now why). I think in general it would be better if Whisper is not RLP depended. However the PoW hash is still RLP depended so this does not really resolve that...

Nice side effect is that testing time of test_shh_connect.nim dropped ~30 seconds.

By the way, I would perhaps move `generateRandomID ` to somewhere more "common", e.g.  `nim-eth/eth/common/utils.nim`?

kdeme commented:
hmm, more than the usual failure on Mac, probably fixed by adding more margin to the `sleepAsync(messageInterval)` .

zah commented on eth/p2p/rlpx_protocols/whisper_protocol.nim:
```diff
@@ -868,14 +877,43 @@ proc sendP2PMessage*(node: EthereumNode, peerId: NodeId, env: Envelope): bool =
+      bestBitTarget = 1
+    else:
+      bestBitTarget = ceil(log(x, 2)).int
+    (msg.env.nonce, msg.hash) = msg.env.minePow(powTime, bestBitTarget)
```
How much time does the proof-of-work take in practice with typical settings? Should we do something to offload it from the main thread? (it will block all network communications at the moment)

zah replied:
It's also possible to make it run on the main thread in small incremental steps that `yield` control back to the I/O loop after a fixed interval of time.

kdeme replied:
I thought I was using the same default minimum PoW as in geth but apparently not.
Anyway, setting it to 0.2 (default of geth) gives me numbers of around 20-30 ms up to 80 ms with TTL=30 and message of 256 bytes (very quick test).

I do like the `yield` idea. I think I'll have a go on this later.

kdeme replied:
Would the use of `poll()` be accepted in this scenario? I think it is the easiest and cleanest (no need to make proc async). However, perhaps we don't want to promote the usage of `poll()` (and even remove it?) due to issues it can create when used in async proc (recursive poll)?

zah replied:
@cheatfate has opposed recurrent usage of `poll` or `waitFor` in the past. I'll let him comment on this and I'll also mention that we could benefit from an `AsyncChannel` facility provided by Chronos is situations like this where the work can be sent to a background thread.

Is the recently added support for pipes enough to implement `AsyncChannel` quickly?

kdeme replied:
OK, to be clear, the `poll` would not be in an async call and thus not re-entrant (at least that is how I understand it). It is just that it is a very direct usage of the call and perhaps now what we want.

kdeme replied:
Actually, it's a bad idea. Giving it some thought, you don't know how the API will be used, and thus you don't know if it will ever end up in a `async` call (unless you can somehow enforce not to allow this?).

In fact, it will already do in the Whisper RPC implementation (and a bit hidden due to the macro, thus extra dangerous). While it did not create any issue while I was testing this, it probably will bite back at the time you least expect it.

Can we merge this PR as is? I will create an issue to improve this.

kdeme commented:
Changed the default minimum PoW to the one that geth has.
Added some marge to the `messageInterval` sleep to be sure that the messages have arrived at the receiving peer before checking. (this gave many errors on test_shh_connect on macOS).
